project/driver_controls/src/driver_io.c

```diff
@@ -0,0 +1,106 @@
+#include <stddef.h>
+
+#include "gpio_it.h"
+#include "driver_io.h"
+#include "driver_callback.h"
+
+// Arbitrary thresholds for gas pedal
+#define COAST_THRESHOLD 1000
+#define DRIVE_THRESHOLD 3000
+
+typedef struct DriverIOSettings {
+  GPIOAddress address;
+  GPIODir direction;
+  GPIOAltFn alt_function;
+  InterruptEdge edge;
+  DriverIOCallback callback;
+} DriverIOSettings;
+
+// Lookup table to map specific input devices to specific pins
+static DriverIOData s_devices[NUM_DRIVER_IO_PIN] = {
+  [DRIVER_IO_PIN_0] = { .id = DRIVER_IO_POWER_SWITCH, .event = INPUT_EVENT_POWER },
+  [DRIVER_IO_PIN_2] = { .id = DRIVER_IO_DIRECTION_SELECTOR,
+                        .event = INPUT_EVENT_DIRECTION_SELECTOR_DRIVE },
+  [DRIVER_IO_PIN_3] = { .id = DRIVER_IO_DIRECTION_SELECTOR,
+                        .event = INPUT_EVENT_DIRECTION_SELECTOR_REVERSE },
+  [DRIVER_IO_PIN_4] = { .id = DRIVER_IO_CRUISE_CONTROL, .event = INPUT_EVENT_CRUISE_CONTROL },
+  [DRIVER_IO_PIN_5] = { .id = DRIVER_IO_CRUISE_CONTROL_INC,
+                        .event = INPUT_EVENT_CRUISE_CONTROL_INC },
+  [DRIVER_IO_PIN_6] = { .id = DRIVER_IO_CRUISE_CONTROL_DEC,
+                        .event = INPUT_EVENT_CRUISE_CONTROL_DEC },
+  [DRIVER_IO_PIN_7] = { .id = DRIVER_IO_TURN_SIGNAL, .event = INPUT_EVENT_TURN_SIGNAL_LEFT },
+  [DRIVER_IO_PIN_8] = { .id = DRIVER_IO_TURN_SIGNAL, .event = INPUT_EVENT_TURN_SIGNAL_RIGHT },
+  [DRIVER_IO_PIN_9] = { .id = DRIVER_IO_HAZARD_LIGHT, .event = INPUT_EVENT_HAZARD_LIGHT },
+  [DRIVER_IO_PIN_10] = { .id = DRIVER_IO_MECHANICAL_BRAKE, .event = INPUT_EVENT_MECHANICAL_BRAKE }
+};
+
+static DriverIOSettings prv_settings(GPIOAddress address, GPIODir dir, InterruptEdge edge,
```

Comment:
This is super inefficient if it isn't optimized by the compiler. You're doing a ton of copies.